### PR TITLE
Add dispersion to reduced diagnostics

### LINKED ISSUE #678: Add beam-based dispersion to reduced diagnostics
Add a beam-based calculation of dispersion to the reduced diagnostics output.

This is especially useful for making comparisons of lattice Twiss properties between multiple codes.

### PR DESCRIPTION
**Pull request by cemitch99**
Modified the reduced beam diagnostics to include beam-based dispersion values `dispersion_x` , `dispersion_y` and their derivatives `dispersion_px`, `dispersion_py`.  Close #678.

The transverse Twiss functions `beta_x`, `beta_y`, `alpha_x`, `alpha_y` were modified to account for correlations with energy in a dispersive beamline.

- [x] Add horizontal dispersion and modify horizontal Twiss calculations
- [x] Add vertical dispersion and modify vertical Twiss calculations
- [x] Modify output of reduced diagnostics to include dispersion
- [x] Add benchmark test
- [x] Add read of reduced diagnostics to analysis script
- [x] Modify documentation

### CODE REVIEW

**Comment by cemitch99** [comment]
[ImpactX_MADX_Twiss_Dogleg_Comparison.pdf](https://github.com/user-attachments/files/16641062/ImpactX_MADX_Twiss_Dogleg_Comparison.pdf)

**Comment by cemitch99** [comment]
All tests currently pass except Windows.  Windows test fails because of trouble opening `diags/reduced_beam_characteristics_final.0.0`

File: examples/dogleg/README.rst [hunk 1]
```diff
@@ -0,0 +1,56 @@
+.. _examples-dogleg:
+
+Dogleg
+=======
```
Comment (ax3l) [inline examples/dogleg/README.rst]:
```suggestion
Dogleg
======
```

File: examples/dogleg/analysis_dogleg.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+
+# initial/final beam
+series = io.Series("diags/openPMD/monitor.h5", io.Access.read_only)
+last_step = list(series.iterations)[-1]
+initial = series.iterations[1].particles["beam"].to_df()
+final = series.iterations[last_step].particles["beam"].to_df()
+
+# final reduced beam diagnostics
+diagnostics_final = pd.read_csv("diags/reduced_beam_characteristics_final.0.0", sep=" ")
```
Comment (ax3l) [inline examples/dogleg/analysis_dogleg.py]:
```suggestion
initial = series.iterations[1].particles["beam"].to_df()
final_beam = series.iterations[last_step].particles["beam"]
final = final_beam.to_df()
```

File: examples/dogleg/analysis_dogleg.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        8.569865e-09,
+    ],
+    rtol=rtol,
+    atol=atol,
+)
+
+print("")
+print("Final Twiss functions:")
+alpha_x, beta_x, alpha_y, beta_y, dispersion_x, dispersion_px = get_twiss(
+    diagnostics_final
```
Comment (ax3l) [inline examples/dogleg/analysis_dogleg.py]:
```suggestion
    final_beam
```

File: examples/dogleg/analysis_dogleg.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    alpha_x = diagnostics_final["alpha_x"].values[0]
+    beta_x = diagnostics_final["beta_x"].values[0]
+    d_x = diagnostics_final["dispersion_x"].values[0]
+    d_px = diagnostics_final["dispersion_px"].values[0]
+    alpha_y = diagnostics_final["alpha_y"].values[0]
+    beta_y = diagnostics_final["beta_y"].values[0]
+    # d_y = diagnostics_final["dispersion_y"].values[0]
+    # d_py = diagnostics_final["dispersion_py"].values[0]
+
+    return (alpha_x, beta_x, alpha_y, beta_y, d_x, d_px)
```
Comment (ax3l) [inline examples/dogleg/analysis_dogleg.py]:
```suggestion
def get_twiss(openpmd_beam):
    """Return Twiss functions from an openPMD particle species

    Returns
    -------
    alpha_x, beta_x, alpha_y, beta_y, dispersion_x, dispersion_px
    """
    alpha_x = openpmd_beam.get_attribute("alpha_x")
    beta_x = openpmd_beam.get_attribute("beta_x")
    d_x = openpmd_beam.get_attribute("dispersion_x")
    d_px = openpmd_beam.get_attribute("dispersion_px")
    alpha_y = openpmd_beam.get_attribute("alpha_y")
    beta_y = openpmd_beam.get_attribute("beta_y")
    # d_y = openpmd_beam.get_attribute("dispersion_y")
    # d_py = openpmd_beam.get_attribute(["dispersion_py")

    return (alpha_x, beta_x, alpha_y, beta_y, d_x, d_px)
```

**Review by ax3l — COMMENTED** [review]
The non-MPI test fails because of the pattern matching of the CSV file (only one `.N` at the end).

We can avoid the CSV file altogether and read directly from the openPMD file we have open like this:

File: examples/dogleg/analysis_dogleg.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    epstrms = beam.cov(ddof=0)
+    emittance_x = (sigx**2 * sigpx**2 - epstrms["position_x"]["momentum_x"] ** 2) ** 0.5
+    emittance_y = (sigy**2 * sigpy**2 - epstrms["position_y"]["momentum_y"] ** 2) ** 0.5
+    emittance_t = (sigt**2 * sigpt**2 - epstrms["position_t"]["momentum_t"] ** 2) ** 0.5
+
+    return (sigx, sigy, sigt, emittance_x, emittance_y, emittance_t)
+
+
+def get_twiss(openpmd_beam):
+    """Return Twiss functions from an openPMD particle species
```
Comment (cemitch99) [inline examples/dogleg/analysis_dogleg.py:37]: I see the changes you made here, but they don't appear as suggestions--my original version is lost.

Comment (ax3l) [reply examples/dogleg/analysis_dogleg.py:37]: I already applied them online :)

You need to `git fetch` and `git pull` them now please.

**Comment by ax3l** [comment]
Note: `Windows / MSVC w/o MPI shared` currently fails #650 but in a later test.

File: examples/dogleg/analysis_dogleg.py (hunk 5 above)
Comment (cemitch99) [reply examples/dogleg/analysis_dogleg.py:37]: Ok, looks great.  I had forgotten that the reduced diagnostics data is also duplicated in the openPMD output.

Comment (ax3l) [reply examples/dogleg/analysis_dogleg.py:37]: yes, that's still quite recent #584

**Comment by ax3l** [comment]
@cemitch99 oh, we forgot to update the documentation in https://impactx.readthedocs.io/en/latest/dataanalysis/dataanalysis.html#reduced-beam-characteristics

**Comment by ax3l** [comment]
Added in #686

**Pull request merged**